File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (!list1 && list2) { return list2; } 
        else if (list1 && !list2) { return list1; }
        else if (!list1 && !list2) {return nullptr; } 

        ListNode* a{list1};
        ListNode* b(list2);
        ListNode* c = (a->val <= b->val) ? a : b;
        if (a->val <= b->val) { a = a->next; }
        else { b = b->next; }

        ListNode* start{c};
        while (a && b) {
            if (a->val <= b->val) {
                c->next = a;
                c = c->next;
                a = a->next;
            }
            else {
                c->next = b;
                c = c->next;
                b = b->next;
            }
        }

        while (a) {
            c->next = a;
            c = c->next;
            a = a->next;
        }

        while (b) {
            c->next = b;
            c = c->next;
            b = b->next;
        }

        return start;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # list of length 0 or 1
        if head == None or head.next == None:
            return

        st = []
        trav = head
        n = 0 # length of list
        while trav != None:
            st.append(trav)
            trav = trav.next
            n += 1

        add_from_stack = True
        i = 1
        trav = head
        nxt = head.next

        while i < n:
            if add_from_stack:
                top = st.pop() # should be fine to call bc never going to get past half
                trav.next = top
            elif nxt:
                trav.next = nxt
                nxt = nxt.next
            add_from_stack ^= True # light switch!
            i += 1
            trav = trav.next

        trav.next = None


            

        
        
        